File: Software/RF2DFieldSolver/CustomWidgets/siunitedit.cpp
#include "siunitedit.h"

#include <System.Windows.Forms.h>
#include <System.ComponentModel.h>
#include <cmath>

SIUnitEdit::SIUnitEdit(System::String^ unit, System::String^ prefixes, int precision, System::Windows::Forms::Form^ parent)
    : TextBox()
{
    this->unit = unit;
    this->prefixes = prefixes;
    this->precision = precision;
    this->TextAlign = System::Windows::Forms::HorizontalAlignment::Center;
    this->Parent = parent;
    this->LostFocus += gcnew System::EventHandler(this, &SIUnitEdit::OnLostFocus);
    this->KeyPress += gcnew System::Windows::Forms::KeyPressEventHandler(this, &SIUnitEdit::OnKeyPress);
    this->KeyDown += gcnew System::Windows::Forms::KeyEventHandler(this, &SIUnitEdit::OnKeyDown);
    this->MouseWheel += gcnew System::Windows::Forms::MouseEventHandler(this, &SIUnitEdit::OnMouseWheel);
    this->TextChanged += gcnew System::EventHandler(this, &SIUnitEdit::OnTextChanged);
    setValueQuiet(0);
}

SIUnitEdit::SIUnitEdit(System::Windows::Forms::Form^ parent)
    : SIUnitEdit("", " ", 4, parent)
{
}

void SIUnitEdit::setValue(double value)
{
    if(value != _value) {
        setValueQuiet(value);
        ValueChanged(this, System::EventArgs::Empty);
        ValueUpdated(this, System::EventArgs::Empty);
    }
}

static char swapUpperLower(char c) {
    if(isupper(c)) {
        return tolower(c);
    } else if(islower(c)) {
        return toupper(c);
    } else {
        return c;
    }
}

void SIUnitEdit::OnKeyPress(System::Object^ sender, System::Windows::Forms::KeyPressEventArgs^ e)
{
    if (e->KeyChar == (char)System::Windows::Forms::Keys::Escape) {
        // abort editing process and set old value
        this->Clear();
        setValueQuiet(_value);
        EditingAborted(this, System::EventArgs::Empty);
        this->Parent->Focus();
        e->Handled = true;
    }
    else if (e->KeyChar == (char)System::Windows::Forms::Keys::Return) {
        // use new value without prefix
        parseNewValue(1.0);
        continueEditing();
        e->Handled = true;
    }
    else if (prefixes->IndexOf(e->KeyChar) >= 0) {
        // a valid prefix key was pressed
        parseNewValue(Unit::SIPrefixToFactor(e->KeyChar));
        continueEditing();
        e->Handled = true;
    }
    else if (prefixes->IndexOf(swapUpperLower(e->KeyChar)) >= 0) {
        // no match on the pressed case but on the upper/lower case instead -> also accept this
        parseNewValue(Unit::SIPrefixToFactor(swapUpperLower(e->KeyChar)));
        continueEditing();
        e->Handled = true;
    }
}

void SIUnitEdit::OnKeyDown(System::Object^ sender, System::Windows::Forms::KeyEventArgs^ e)
{
    if (e->KeyCode == System::Windows::Forms::Keys::Escape) {
        // abort editing process and set old value
        this->Clear();
        setValueQuiet(_value);
        EditingAborted(this, System::EventArgs::Empty);
        this->Parent->Focus();
        e->Handled = true;
    }
    else if (e->KeyCode == System::Windows::Forms::Keys::Return) {
        // use new value without prefix
        parseNewValue(1.0);
        continueEditing();
        e->Handled = true;
    }
}

void SIUnitEdit::OnMouseWheel(System::Object^ sender, System::Windows::Forms::MouseEventArgs^ e)
{
    if(_value == 0.0) {
        // can't figure out step size with zero value
        return;
    }
    // most mousewheel have 15 degree increments, the reported delta is in 1/8th degree -> 120
    auto increment = e->Delta / 120.0;
    // round toward bigger step in case of special higher resolution mousewheel
    unsigned int steps = std::abs(increment > 0 ? ceil(increment) : floor(increment));
    int sign = increment > 0 ? 1 : -1;
    // figure out step increment
    auto newVal = _value;
    if(this->Focused) {
        auto cursor = this->SelectionStart;
        if(cursor == 0) {
            // cursor in front of first digit, do nothing (too big of a change, probably over/underflows)
            return;
        }
        // change the digit at the current cursor
        int nthDigit = cursor;
        // account for decimal point/leading zero/sign
        if(_value < 0) {
            nthDigit--;
        }
        auto dotPos = this->Text->IndexOf('.');
        if(dotPos >= 0 && dotPos < nthDigit) {
            nthDigit--;
        }
        if(this->Text->StartsWith("-0.") || this->Text->StartsWith("0.")) {
            nthDigit--;
        }
        auto step_size = pow(10, floor(log10(std::abs(newVal))) - nthDigit + 1);
        newVal += step_size * steps * sign;
        setValue(newVal);
        this->Text = this->PlaceholderText;
        this->SelectionStart = cursor;
    } else {
        // default to the third digit
        constexpr int nthDigit = 3;
        while(steps > 0) {
            // do update in multiple steps because the step size could change inbetween
            auto step_size = pow(10, floor(log10(std::abs(newVal))) - nthDigit + 1);
            newVal += step_size * sign;
            steps--;
        }
        setValue(newVal);
    }
}

void SIUnitEdit::OnLostFocus(System::Object^ sender, System::EventArgs^ e)
{
    parseNewValue(1.0);
    FocusLost(this, System::EventArgs::Empty);
}

void SIUnitEdit::OnTextChanged(System::Object^ sender, System::EventArgs^ e)
{
    parseNewValue(1.0);
}

void SIUnitEdit::setValueQuiet(double value)
{
    _value = value;
    this->PlaceholderText = Unit::ToString(value, unit, prefixes, precision);
    if(!this->Text->IsEmpty()) {
        // currently editing, update the text as well
        continueEditing();
    }
}

void SIUnitEdit::parseNewValue(double factor)
{
    System::String^ input = this->Text;
    if(input->IsEmpty()) {
        setValueQuiet(_value);
        EditingAborted(this, System::EventArgs::Empty);
    } else {
        // remove optional unit
        if(input->EndsWith(unit)) {
            input = input->Substring(0, input->Length - unit->Length);
        }
        auto lastChar = input[input->Length - 1];
        if(prefixes->IndexOf(lastChar) >= 0) {
            factor = Unit::SIPrefixToFactor(lastChar);
            input = input->Substring(0, input->Length - 1);
        }
        // remaining input should only contain numbers
        double v;
        if(Double::TryParse(input, v)) {
            this->Clear();
            setValue(v * factor);
        } else {
            System::Diagnostics::Debug::WriteLine("SIUnit conversion failure: " + input);
        }
    }
}

void SIUnitEdit::continueEditing()
{
    this->Text = this->PlaceholderText;
    this->SelectAll();
}

File: Software/RF2DFieldSolver/CustomWidgets/SiUnitEdit.cs
using System;
using System.Windows.Forms;

namespace RF2DFieldSolver.CustomWidgets
{
    public class SiUnitEdit : TextBox
    {
        private string unit;
        private string prefixes;
        private int precision;
        private double value;

        public SiUnitEdit(string unit = "", string prefixes = " ", int precision = 0)
        {
            this.unit = unit;
            this.prefixes = prefixes;
            this.precision = precision;
            this.TextAlign = HorizontalAlignment.Center;
            this.Leave += (sender, e) => ParseNewValue(1.0);
            SetValueQuiet(0);
        }

        public void SetUnit(string unit)
        {
            this.unit = unit;
            SetValueQuiet(value);
        }

        public void SetPrefixes(string prefixes)
        {
            this.prefixes = prefixes;
            SetValueQuiet(value);
        }

        public void SetPrecision(int precision)
        {
            this.precision = precision;
            SetValueQuiet(value);
        }

        public double GetValue()
        {
            return value;
        }

        public void SetValue(double value)
        {
            if (value != this.value)
            {
                SetValueQuiet(value);
                OnValueChanged(EventArgs.Empty);
            }
        }

        protected override void OnKeyPress(KeyPressEventArgs e)
        {
            base.OnKeyPress(e);
            if (e.KeyChar == (char)Keys.Escape)
            {
                Clear();
                SetValueQuiet(value);
                OnEditingAborted(EventArgs.Empty);
                e.Handled = true;
            }
            else if (e.KeyChar == (char)Keys.Return)
            {
                ParseNewValue(1.0);
                ContinueEditing();
                e.Handled = true;
            }
            else if (prefixes.IndexOf(e.KeyChar) >= 0)
            {
                ParseNewValue(SiPrefixToFactor(e.KeyChar));
                ContinueEditing();
                e.Handled = true;
            }
        }

        protected override void OnMouseWheel(MouseEventArgs e)
        {
            base.OnMouseWheel(e);
            if (value == 0.0)
            {
                return;
            }

            int increment = e.Delta / 120;
            int steps = Math.Abs(increment);
            int sign = increment > 0 ? 1 : -1;
            double newVal = value;

            if (Focused)
            {
                int cursor = SelectionStart;
                if (cursor == 0)
                {
                    return;
                }

                int nthDigit = cursor;
                if (value < 0)
                {
                    nthDigit--;
                }

                int dotPos = Text.IndexOf('.');
                if (dotPos >= 0 && dotPos < nthDigit)
                {
                    nthDigit--;
                }

                if (Text.StartsWith("-0.") || Text.StartsWith("0."))
                {
                    nthDigit--;
                }

                double stepSize = Math.Pow(10, Math.Floor(Math.Log10(Math.Abs(newVal))) - nthDigit + 1);
                newVal += stepSize * steps * sign;
                SetValue(newVal);
                Text = PlaceholderText;
                SelectionStart = cursor;
            }
            else
            {
                const int nthDigit = 3;
                while (steps > 0)
                {
                    double stepSize = Math.Pow(10, Math.Floor(Math.Log10(Math.Abs(newVal))) - nthDigit + 1);
                    newVal += stepSize * sign;
                    steps--;
                }
                SetValue(newVal);
            }
        }

        private void SetValueQuiet(double value)
        {
            this.value = value;
            PlaceholderText = ToString(value, unit, prefixes, precision);
            if (!string.IsNullOrEmpty(Text))
            {
                ContinueEditing();
            }
        }

        private void ParseNewValue(double factor)
        {
            string input = Text;
            if (string.IsNullOrEmpty(input))
            {
                SetValueQuiet(value);
                OnEditingAborted(EventArgs.Empty);
            }
            else
            {
                if (input.EndsWith(unit))
                {
                    input = input.Substring(0, input.Length - unit.Length);
                }

                char lastChar = input[input.Length - 1];
                if (prefixes.IndexOf(lastChar) >= 0)
                {
                    factor = SiPrefixToFactor(lastChar);
                    input = input.Substring(0, input.Length - 1);
                }

                if (double.TryParse(input, out double parsedValue))
                {
                    Clear();
                    SetValue(parsedValue * factor);
                }
                else
                {
                    Console.WriteLine("SIUnit conversion failure: " + input);
                }
            }
        }

        private void ContinueEditing()
        {
            Text = PlaceholderText;
            SelectAll();
        }

        private double SiPrefixToFactor(char prefix)
        {
            switch (prefix)
            {
                case 'p': return 1e-12;
                case 'n': return 1e-9;
                case 'u': return 1e-6;
                case 'm': return 1e-3;
                case 'k': return 1e3;
                case 'M': return 1e6;
                case 'G': return 1e9;
                default: return 1.0;
            }
        }

        private string ToString(double value, string unit, string prefixes, int precision)
        {
            string format = "F" + precision;
            return value.ToString(format) + unit;
        }

        protected virtual void OnValueChanged(EventArgs e)
        {
            ValueChanged?.Invoke(this, e);
        }

        protected virtual void OnEditingAborted(EventArgs e)
        {
            EditingAborted?.Invoke(this, e);
        }

        public event EventHandler ValueChanged;
        public event EventHandler EditingAborted;
    }
}
